### ISSUE TITLE
JPEG XL

### ISSUE THREAD

**Issue by ondrs** [imgproxy/imgproxy#922]
Hello,

do you plan to add support for JPEG XL format in imgproxy?
It is supported libvips and in Chromium and Firefox nightly.

**Comment by nikeee** [comment]
Update: JXL is currently not supported any more by Chrome and Firefox, but Safari will ship with support for it. The situation might change with Firefox/Chrome.

**Comment by obrhoff** [comment]
another vote, especially since apple announced system wide support for iOS17 / macOS14.

**Comment by ondrs** [comment]
Safari 17 with JPEG XL support has been shipped
https://webkit.org/blog/14445/webkit-features-in-safari-17-0/

**Comment by obrhoff** [comment]
Another bump. I would really love to move to JPEG XL for our iOS app. Keep in mind that JPEG XL supports lossless transcoding from JPEG to JPEG XL.

**Comment by DarthSim** [comment]
Hey everyone!

We're working on JPEG XL support: https://github.com/libvips/libvips/pull/3712.

Be aware though that the JPEG XL encoder needs a lot of memory: https://twitter.com/darth_sim/status/1711794478778814844

**Comment by chendo** [comment]
@DarthSim I see that libvips PR has been merged; does that mean imgproxy should support JPEG XL soon?

**Comment by DarthSim** [comment]
@chendo The PR was merged to vips 8.16 which is not released yet.

And honestly, I don't quite understand why you are still hyped up about JXL. It is still supported only by Safari and it still requires an enormous amount of RAM for encoding. So nowadays it looks like a good way to shoot yourself in the foot for a single browser's sake.

**Comment by sbaildon** [comment]
> @chendo The PR was merged to vips 8.16 which is not released yet.
> 
> And honestly, I don't quite understand why you are still hyped up about JXL. It is still supported only by Safari and it still requires an enormous amount of RAM for encoding. So nowadays it looks like a good way to shoot yourself in the foot for a single browser's sake.

regarding the memory comment — is that still true? see https://github.com/libjxl/libjxl/issues/1106#issuecomment-2060730300

**Comment by DarthSim** [comment]
@sbaildon They indeed improved things, and the compression of a 7360x4912 image now consumes 664M instead of 1.8G, but this is still huge. On the other hand, compression to AVIF consumes more than that.

**Comment by chendo** [comment]
@DarthSim ahh, I see. In our case, we're mostly wanting JXL decoding support so we can have our source files be smaller and thus take less time for imgproxy to fetch. We're not looking to implement JXL for end users just yet, but our analytics indicate about 50% of our users are on Safari versions that support JXL.

**Comment by DarthSim** [comment]
@chendo Nah, the case is that libvips with proper JXL support has not been released yet. The changes from my PR should go to v8.16 but it seems like it's delayed. I sometimes do patch vips right in the Docker image but I don't think this is a good idea in this exact case.

**Comment by jonnyawsom3** [comment]
Libvips 8.16 released yesterday with animation and much better metadata support (ICC profiles, EXIF, ect). The only thing missing is HDR although that may not be required at this time anyway

**Comment by DarthSim** [comment]
JPEG XL support was added to the `latest` build

**Comment by obrhoff** [comment]
Just gave it a try and works so far great in iOS. Thanks! ❤️

**Comment by jonnyawsom3** [comment]
This might not be the right place to say it, but I'd recommend setting the default effort to 5 instead 4, which enables the Variable blocks of VarDCT. 4 and below use fixed 8x8 like the old JPEG. If possible, you could also add a separate effort setting for lossless, to maintain speed for both use cases.

**Comment by DarthSim** [comment]
@jonnyawsom3 Effort 5 dramatically slows down encoding compared to effort 4. We did some tests with SSIMULACRA2 and found that effort 4 still shows surprisingly good results. It's not as good as effort 5+, but still good.

When we added AVIF support, the most common issue here on GH was "imgproxy saves to AVIF too slowly" or "imgproxy times out when I use AVIF." I had to reply to every single one of them that this is because AVIF is slow, we can do nothing about this, but here's the config that you can use to change the speed, etc. In the end, I gave up and increased the default AVIF speed.

Considering that most people will adopt JPEG XL just because it's a new "best in the world" image format and because Apple added its support to Safari, the situation will be exactly the same if not worse.

So, we tried to find a sweet spot between good encoding speed and good results. Our defaults should work well for those who just want JPEG XL. For the rest, the `IMGPROXY_JXL_EFFORT` config is at their service.
